docs/concepts/chain-key-design.md

```diff
@@ -0,0 +1,211 @@
+
+// This works - querying by gameId, seasonId, playerId, itemId, and timestamp
+const playerItemSeasonTimestampItems = await getObjectsByPartialCompositeKey(
+  ctx,
+  GameItem.INDEX_KEY,
+  ["game123", "season456", "player789", "item987", "timestamp123"],
+  GameItem
+);
+
+For organizing collections of related objects:
+
+```typescript
+export class UserAsset extends ChainObject {
+  @ChainKey({ position: 0 }) public collection: string = 'user-assets';
+  @ChainKey({ position: 1 }) public userId: string;
+  @ChainKey({ position: 2 }) public assetId: string;
+  
+  // Asset properties
+  public quantity: number;
+}
+
+// Usage:
+const allUserAssets = await ctx.store.query(UserAsset, { 
+  collection: 'user-assets' 
+});
+const userAssets = await ctx.store.query(UserAsset, { 
+  collection: 'user-assets', 
+  userId: 'user123' 
+});
+```
+
+### Time-Based Organization
+
+For time-series data or events:
+
+```typescript
+export class GameEvent extends ChainObject {
+  @ChainKey({ position: 0 }) public eventType: string;
+  @ChainKey({ position: 1 }) public year: number;
+  @ChainKey({ position: 2 }) public month: number;
+  @ChainKey({ position: 3 }) public day: number;
+  @ChainKey({ position: 4 }) public eventId: string;
+  
+  // Event data
+  public data: any;
+}
+
+// Query all login events from October 2023
+const octoberLogins = await ctx.store.query(GameEvent, {
+  eventType: 'login',
+  year: 2023,
+  month: 10
+});
+```
+
+### Hierarchical Data
+
+For parent-child relationships:
+
+```typescript
+export class NestedItem extends ChainObject {
+  @ChainKey({ position: 0 }) public readonly rootType: string;
+  @ChainKey({ position: 1 }) public readonly parentId: string | null;
+  @ChainKey({ position: 2 }) public readonly itemId: string;
+  
+  // Item data
+  public name: string;
+  public data: any;
+}
+
+// Query all top-level items (no parent)
+const rootItems = await ctx.store.query(NestedItem, {
+  rootType: 'folder',
+  parentId: null
+});
+
+// Query all children of a specific parent
+const childItems = await ctx.store.query(NestedItem, {
+  rootType: 'folder',
+  parentId: 'parent123'
+});
+```
+
+## Common Pitfalls to Avoid
+
+1. **Skipping Key Indexes**: Always use sequential numbers starting with 0.
+   ```typescript
+   // Incorrect
+   @ChainKey({position: 0}) public readonly type: string;
+   @ChainKey({position: 2}) public readonly id: string; // Gap will cause issues
+   
+   // Correct
+   @ChainKey({position: 0}) public readonly type: string;
+   @ChainKey({position: 1}) public readonly id: string;
+   ```
+   
+   
+2. **Assigning Mutable properties as Chain Key Values**: Avoid using properties that change as chain keys. Changing a chain key would effectively represent a new entity.
+
+4. **Using Non-Deterministic Values**: Timestamps or random values make poor chain keys. However, placing values like these at the end of a sequence of chain keys can be a useful strategy to ensure data integrity and prevent collisions.
```

Comment:
I know that's not the case here, but that's a tricky thing to mention "random" and "timestamp" words here. Maybe it deserves a mention that any non-deterministic values cannot be created by the chaincode.